### PR TITLE
[Merged by Bors] - feat: uniqueness of integral curves of a vector field on a manifold

### PR DESCRIPTION
**Pull request by winstonyin**
We prove the uniqueness of integral curves of a vector field on a manifold using the uniqueness theorem for solutions to ODEs.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by grunweg — COMMENTED** [review]
I couldn't resist and took a look already. I figured you'll still clean up your uniqueness proofs, hence I didn't comment on these yet.

File: Mathlib/Analysis/ODE/Gronwall.lean [hunk 1]
```diff
@@ -223,25 +226,83 @@ theorem dist_le_of_trajectories_ODE {v : ℝ → E → E} {K : ℝ≥0} (hv : 
...
 
+-- Ico rather than Icc
+theorem ODE_solution_unique_of_mem_set_Ico {v : ℝ → E → E} {s : ℝ → Set E} {K : ℝ≥0}
+    (hv : ∀ t, LipschitzOnWith K (v t) (s t))
+    {f g : ℝ → E} {a b : ℝ} (hf : ContinuousOn f (Ico a b))
+    (hf' : ∀ t ∈ Ico a b, HasDerivWithinAt f (v t (f t)) (Ici t) t) (hfs : ∀ t ∈ Ico a b, f t ∈ s t)
+    (hg : ContinuousOn g (Ico a b)) (hg' : ∀ t ∈ Ico a b, HasDerivWithinAt g (v t (g t)) (Ici t) t)
+    (hgs : ∀ t ∈ Ico a b, g t ∈ s t) (ha : f a = g a) : EqOn f g (Ico a b) := fun _ ht ↦
+  ODE_solution_unique_of_mem_set hv
+    (hf.mono (Icc_subset_Ico_right ht.2)) (fun _ ht' ↦ hf' _ ⟨ht'.1, lt_trans ht'.2 ht.2⟩)
```
[the remark this answers is not in the document]
Comment (winstonyin) [reply Mathlib/Analysis/ODE/Gronwall.lean]: I collected these repetitive phrases into `have`s. But yes, a tactic call would be ideal.

File: Mathlib/Analysis/ODE/Gronwall.lean [hunk 2]
```diff
@@ -223,25 +223,91 @@ theorem dist_le_of_trajectories_ODE {v : ℝ → E → E} {K : ℝ≥0} (hv : 
...
+        simp
+      · intros t' ht'
+        apply HasDerivAt.hasDerivWithinAt
+        convert HasFDerivAt.comp_hasDerivAt t' (hg' (-t') (hmt ht')) (hasDerivAt_neg t')
+        simp
+    rw [eqOn_comp_right_iff] at this
+    convert this
+    simp
+  · -- case `t ≥ t₀`: follows trivially from the `Ico` version of the uniqueness lemma
+    have hss : Ico t₀ b ⊆ Ioo a b := fun _ ht' ↦ mem_of_mem_of_subset ht' (Ico_subset_Ioo_left ht.1)
```
Comment (grunweg) [inline Mathlib/Analysis/ODE/Gronwall.lean:302]: This would be a good lemma to Mathlib.Data.Set.Intervals.Basic. I presume whoever reviews this will also ask you to add several dual versions :-)

File: Mathlib/Analysis/ODE/Gronwall.lean [hunk 3]
```diff
@@ -250,8 +316,7 @@ theorem ODE_solution_unique {v : ℝ → E → E} {K : ℝ≥0} (hv : ∀ t, Lip
     (hg' : ∀ t ∈ Ico a b, HasDerivWithinAt g (v t (g t)) (Ici t) t) (ha : f a = g a) :
     ∀ t ∈ Icc a b, f t = g t :=
```
Comment (grunweg) [inline Mathlib/Analysis/ODE/Gronwall.lean:317]: Pre-existing, but while you're at it:
```suggestion
    EqOn f g (Icc a b) :=

File: Mathlib/Analysis/ODE/Gronwall.lean [hunk 4]
```diff
@@ -223,25 +223,91 @@ theorem dist_le_of_trajectories_ODE {v : ℝ → E → E} {K : ℝ≥0} (hv : 
...
+    (hgs : ∀ t ∈ Ioo a b, g t ∈ s t) (ha : f t₀ = g t₀) : EqOn f g (Ioo a b) := by
+  -- split `Ioo a b = Ioc a t₀ ∪ Ico t₀ b` and treat the two cases separately
+  apply EqOn.mono (Ioo_subset_Ioc_union_Ico (b := t₀))
+  apply EqOn.union
+  · -- case `t ≤ t₀`: show `fun t ↦ f_or_g (-t)` solves `-v (-t)` within `s (-t)` by composition,
+    -- so we can use the `Ico` version of the uniqueness lemma backwards in time
+    have hv' : ∀ t, LipschitzOnWith K (Neg.neg ∘ (v (-t))) (s (-t)) := fun t ↦ by
+      rw [← one_mul K]
+      exact LipschitzWith.id.neg.comp_lipschitzOnWith (hv _)
+    have : EqOn (f ∘ Neg.neg) (g ∘ Neg.neg) (Ico (-t₀) (-a)) := by
```
Comment (grunweg) [inline Mathlib/Analysis/ODE/Gronwall.lean:284]: In any case, if there's no nice tactic for this, this is a useful general lemma.

File: Mathlib/Analysis/ODE/Gronwall.lean [hunk 5]
```diff
@@ -223,25 +223,91 @@ theorem dist_le_of_trajectories_ODE {v : ℝ → E → E} {K : ℝ≥0} (hv : 
...
     (hg : ContinuousOn g (Icc a b)) (hg' : ∀ t ∈ Ico a b, HasDerivWithinAt g (v t (g t)) (Ici t) t)
-    (hgs : ∀ t ∈ Ico a b, g t ∈ s t) (ha : f a = g a) : ∀ t ∈ Icc a b, f t = g t := fun t ht ↦ by
+    (hgs : ∀ t ∈ Ico a b, g t ∈ s t) (ha : f a = g a) : EqOn f g (Icc a b) := fun t ht ↦ by
   have := dist_le_of_trajectories_ODE_of_mem_set hv hf hf' hfs hg hg' hgs (dist_le_zero.2 ha) t ht
   rwa [zero_mul, dist_le_zero] at this
 set_option linter.uppercaseLean3 false in
 #align ODE_solution_unique_of_mem_set ODE_solution_unique_of_mem_set
 
+/-- There exists only one solution of an ODE \(\dot x=v(t, x)\) in a set `s ⊆ ℝ × E` with
+a given initial value provided that RHS is Lipschitz continuous in `x` within `s`,
```
Comment (grunweg) [inline Mathlib/Analysis/ODE/Gronwall.lean:248]: Nit, partially pre-existing: in my book, "provided that the RHS" or "provided the RHS" would be correct grammar.

File: Mathlib/Analysis/ODE/Gronwall.lean [hunk 6]
```diff
@@ -223,25 +223,91 @@ theorem dist_le_of_trajectories_ODE {v : ℝ → E → E} {K : ℝ≥0} (hv : 
...
+and we consider only solutions included in `s`.
+
+This version shows uniqueness in a half-open interval `Ico a b`, where `a` is the initial time. -/
+theorem ODE_solution_unique_of_mem_set_Ico {v : ℝ → E → E} {s : ℝ → Set E} {K : ℝ≥0}
+    (hv : ∀ t, LipschitzOnWith K (v t) (s t))
+    {f g : ℝ → E} {a b : ℝ} (hf : ContinuousOn f (Ico a b))
+    (hf' : ∀ t ∈ Ico a b, HasDerivWithinAt f (v t (f t)) (Ici t) t) (hfs : ∀ t ∈ Ico a b, f t ∈ s t)
+    (hg : ContinuousOn g (Ico a b)) (hg' : ∀ t ∈ Ico a b, HasDerivWithinAt g (v t (g t)) (Ici t) t)
+    (hgs : ∀ t ∈ Ico a b, g t ∈ s t) (ha : f a = g a) : EqOn f g (Ico a b) := fun _ ht ↦
```
Comment (grunweg) [inline Mathlib/Analysis/ODE/Gronwall.lean:257]: I think it'd be better style to write the conclusion on a separate line. (I'm not very sure if this is consensus among maintainers.)

The same comment applies to the lemmas immediately above and below.

File: Mathlib/Analysis/ODE/Gronwall.lean [hunk 7]
```diff
@@ -223,25 +223,91 @@ theorem dist_le_of_trajectories_ODE {v : ℝ → E → E} {K : ℝ≥0} (hv : 
...
+    {f g : ℝ → E} {a b t₀ : ℝ} (ht : t₀ ∈ Ioo a b) (hf : ContinuousOn f (Ioo a b))
+    (hf' : ∀ t ∈ Ioo a b, HasDerivAt f (v t (f t)) t) (hfs : ∀ t ∈ Ioo a b, f t ∈ s t)
+    (hg : ContinuousOn g (Ioo a b)) (hg' : ∀ t ∈ Ioo a b, HasDerivAt g (v t (g t)) t)
+    (hgs : ∀ t ∈ Ioo a b, g t ∈ s t) (ha : f t₀ = g t₀) : EqOn f g (Ioo a b) := by
+  -- split `Ioo a b = Ioc a t₀ ∪ Ico t₀ b` and treat the two cases separately
+  apply EqOn.mono (Ioo_subset_Ioc_union_Ico (b := t₀))
+  apply EqOn.union
+  · -- case `t ≤ t₀`: show `fun t ↦ f_or_g (-t)` solves `-v (-t)` within `s (-t)` by composition,
+    -- so we can use the `Ico` version of the uniqueness lemma backwards in time
+    have hv' : ∀ t, LipschitzOnWith K (Neg.neg ∘ (v (-t))) (s (-t)) := fun t ↦ by
```
Comment (grunweg) [inline Mathlib/Analysis/ODE/Gronwall.lean:281]: By the way, you can use parameters in have:
```suggestion
    have hv' (t) : LipschitzOnWith K (Neg.neg ∘ (v (-t))) (s (-t)) := by

**Review by grunweg — COMMENTED** [review]
I left some more comments on the Gronwall estimates.
These applications of `ODE_solution_unique_of_mem_set` feel repetitive, but I don't see a direct way to avoid these.

File: Mathlib/Analysis/ODE/Gronwall.lean [hunk 8]
```diff
@@ -150,8 +150,8 @@ people call this Grönwall's inequality too.
 
 This version assumes all inequalities to be true in some time-dependent set `s t`,
 and assumes that the solutions never leave this set. -/
-theorem dist_le_of_approx_trajectories_ODE_of_mem_set {v : ℝ → E → E} {s : ℝ → Set E} {K : ℝ}
-    (hv : ∀ t, ∀ᵉ (x ∈ s t) (y ∈ s t), dist (v t x) (v t y) ≤ K * dist x y)
+theorem dist_le_of_approx_trajectories_ODE_of_mem_set {v : ℝ → E → E} {s : ℝ → Set E} {K : ℝ≥0}
```
[the remark this answers is not in the document]
Comment (winstonyin) [reply Mathlib/Analysis/ODE/Gronwall.lean:153]: Split out to #8920

**Comment by winstonyin** [comment]
I split out an aux lemma, so the same things don't have to be proven twice for `γ` and `γ'`. A lot of the duplicate code has to do with the nature of "epsilon-shrinking" proofs. You get a handful of `ε_1, ..., ε_n` that each correspond to a statement `P_1 ε_1, ..., P_n ε_n` in the current context as hypotheses. Suppose each `P_i ε` each still holds for all `ε ≤ ε_i`. You then form the minimum of all the `ε_i`s to use for your final statement.

Currently, you have to individually prove `P_i ε_i → ε ≤ ε_i → P_i ε`, interrupting the flow of the overall theorem. I suggest these should be synthesised from a handful of monotonicity lemmas stated elsewhere. Being able to epsilon-shrinking easier will significantly improve the experience of working in analysis.

File: Mathlib/Geometry/Manifold/IntegralCurve.lean [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+  have haux := hnhds.and <| hsrc.and <| hγ.and <| hnhds'.and <| hsrc'.and hγ'
+  rw [Metric.eventually_nhds_iff_ball] at haux
+
+  obtain ⟨ε, hε, haux⟩ := haux
+  refine ⟨ε, hε, ?_⟩
+
+  -- break out all the conditions again
+  have hmem := fun t ht => mem_preimage.mp <| mem_of_mem_nhds (haux t ht).1
+  have hsrc := fun t ht => mem_preimage.mp <| mem_of_mem_nhds (haux t ht).2.1
+  have hmfd : IsIntegralCurveOn _ _ _ := fun t ht => (haux t ht).2.2.1
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/IntegralCurve.lean:413]: `hmfd` is not very clear to me as an outside person. Is `hcurve` a better name?

File: Mathlib/Geometry/Manifold/IntegralCurve.lean [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+  have hmem' := fun t ht => mem_preimage.mp <| mem_of_mem_nhds (haux t ht).2.2.2.1
+  have hsrc' := fun t ht => mem_preimage.mp <| mem_of_mem_nhds (haux t ht).2.2.2.2.1
+  have hmfd' : IsIntegralCurveOn _ _ _ := fun t ht => (haux t ht).2.2.2.2.2
+
+  -- `γ` and `γ'` when expressed in the local chart are continuous on this neighbourhood
+  have hcont := (continuousOn_extChartAt I (γ t₀)).comp
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/IntegralCurve.lean:419]: You can use dot notation again.
`have hcont := (continuousOn_extChartAt I (γ t₀)).comp hmfd.continuousOn hsrc`

File: Mathlib/Geometry/Manifold/IntegralCurve.lean [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+
+  If a continuously differentiable vector field `v` admits two global integral curves
+  `γ γ' : ℝ → M`, and `γ t₀ = γ' t₀` for some `t₀`, then `γ` and `γ'` are equal. -/
+theorem isIntegralCurve_eq_of_contMDiff (hγt : ∀ t, I.IsInteriorPoint (γ t))
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M)))
+    (hγ : IsIntegralCurve γ v) (hγ' : IsIntegralCurve γ' v) (h : γ t₀ = γ' t₀) : γ = γ' := by
+  ext t
+  obtain ⟨T, hT, ht⟩ : ∃ T > 0, t ∈ Ioo (t₀ - T) (t₀ + T) := by
+    refine ⟨2 * |t - t₀| + 1, add_pos_of_nonneg_of_pos (by simp) zero_lt_one, ?_⟩
+    rw [mem_Ioo]
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/IntegralCurve.lean:536]:
```suggestion

File: Mathlib/Geometry/Manifold/IntegralCurve.lean [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+    by_cases ht : t - t₀ < 0
+    · rw [abs_of_neg ht]
+      constructor <;> linarith
+    · rw [abs_of_nonneg (not_lt.mp ht)]
+      constructor <;> linarith
+  exact isIntegralCurveOn_Ioo_eqOn_of_contMDiff
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/IntegralCurve.lean:542]: Dot notation opportunity.
```lean
  exact isIntegralCurveOn_Ioo_eqOn_of_contMDiff
    (Real.ball_eq_Ioo t₀ T ▸ Metric.mem_ball_self hT) (fun t _ => hγt t) hv
    ((hγ.isIntegralCurveOn _).mono (subset_univ _))
    ((hγ'.isIntegralCurveOn _).mono (subset_univ _)) h ht`

File: Mathlib/Geometry/Manifold/IntegralCurve.lean [hunk 13]
```diff
@@ -0,0 +1,551 @@
...
+
+/-- Global integral curves are unique.
+
+  If a continuously differentiable vector field `v` admits two global integral curves
+  `γ γ' : ℝ → M`, and `γ t₀ = γ' t₀` for some `t₀`, then `γ` and `γ'` are equal. -/
+theorem isIntegralCurve_eq_of_contMDiff (hγt : ∀ t, I.IsInteriorPoint (γ t))
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M)))
+    (hγ : IsIntegralCurve γ v) (hγ' : IsIntegralCurve γ' v) (h : γ t₀ = γ' t₀) : γ = γ' := by
+  ext t
+  obtain ⟨T, hT, ht⟩ : ∃ T > 0, t ∈ Ioo (t₀ - T) (t₀ + T) := by
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/IntegralCurve.lean:534]: Can you extract this as a separate lemma?

File: Mathlib/Geometry/Manifold/IntegralCurve.lean [hunk 14]
```diff
@@ -0,0 +1,551 @@
...
+The uniqueness theorem requires the manifold to be Hausdorff (T2), so that the set on which two
+continuous functions agree is closed.
+
+We state simpler versions of the theorem for manifolds without boundary as corollaries.
+
+## To-do
+
+- The case where the integral curve may venture to the boundary of the manifold. See Theorem 9.34,
+  J. M. Lee. May require submanifolds.
+
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/IntegralCurve.lean:51]: Can you add Lee's book as a reference, especially when you're referring to it?

File: Mathlib/Geometry/Manifold/IntegralCurve.lean [hunk 15]
```diff
@@ -0,0 +1,551 @@
...
+  If a continuously differentiable vector field `v` admits two integral curves `γ γ' : ℝ → M`
+  on some open interval `Ioo a b`, and `γ t₀ = γ' t₀` for some `t ∈ Ioo a b`, then `γ` and `γ'`
+  agree on `Ioo a b`. -/
+theorem isIntegralCurveOn_Ioo_eqOn_of_contMDiff (ht₀ : t₀ ∈ Ioo a b)
+    (hγt : ∀ t ∈ Ioo a b, I.IsInteriorPoint (γ t))
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M)))
+    (hγ : IsIntegralCurveOn γ v (Ioo a b)) (hγ' : IsIntegralCurveOn γ' v (Ioo a b))
+    (h : γ t₀ = γ' t₀) : EqOn γ γ' (Ioo a b) := by
+  /-
+  strategy:
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/IntegralCurve.lean:475]: Can you rephrase this? To a casual reader, it's not clear what J is.

File: Mathlib/Geometry/Manifold/IntegralCurve.lean [hunk 16]
```diff
@@ -0,0 +1,551 @@
...
+  * Lee P.213, just need to translate "S is closed in J" to type theory language
+  -/
+  set s := {t | γ t = γ' t} ∩ Ioo a b with hs
+  have hsub : Ioo a b ⊆ s := by
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/IntegralCurve.lean:479]: What do you think about using `suffices` here?
`  suffices h : Ioo a b ⊆ s from fun t ht ↦ mem_setOf.mp ((subset_def ▸ h) t ht).1`,
and then `hsub` becomes your main goal.

File: Mathlib/Geometry/Manifold/IntegralCurve.lean [hunk 17]
```diff
@@ -0,0 +1,551 @@
...
+      ⟨t₀, ⟨ht₀, ⟨h, ht₀⟩⟩⟩
+    · -- is this really the most convenient way to pass to subtype topology?
+      rw [hs, ← Subtype.image_preimage_val, ← Subtype.image_preimage_val,
+        image_subset_image_iff Subtype.val_injective, preimage_setOf_eq]
+      intros t ht
+      rw [mem_preimage, ← closure_subtype] at ht
+      revert ht t
+      apply IsClosed.closure_subset
+      apply isClosed_eq
+      · rw [continuous_iff_continuousAt]
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/IntegralCurve.lean:490]: I also wonder if there's a nicer way to perform this passage. (I don't know one now.)

File: Mathlib/Geometry/Manifold/InteriorBoundary.lean [hunk 18]
```diff
@@ -40,6 +40,8 @@ this requires a definition of submanifolds
 
 open Set
 
+open scoped Manifold
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/InteriorBoundary.lean:43]: Is there a reason for this? It seems to me that you can just remove this.

**Review by grunweg — COMMENTED** [review]
Now that the dependencies have landed or are close to it, I took a look at the main results also. A few, mostly superficial, comments.

File: Mathlib/Geometry/Manifold/IntegralCurve.lean [hunk 19]
```diff
@@ -0,0 +1,551 @@
...
+      constructor <;> linarith
+  exact isIntegralCurveOn_Ioo_eqOn_of_contMDiff
+    (Real.ball_eq_Ioo t₀ T ▸ Metric.mem_ball_self hT) (fun t _ => hγt t) hv
+    (IsIntegralCurveOn.mono (hγ.isIntegralCurveOn _) (subset_univ _))
+    (IsIntegralCurveOn.mono (hγ'.isIntegralCurveOn _) (subset_univ _)) h ht
+
+theorem isIntegralCurve_Ioo_eq_of_contMDiff_boundaryless [BoundarylessManifold I M]
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M)))
+    (hγ : IsIntegralCurve γ v) (hγ' : IsIntegralCurve γ' v) (h : γ t₀ = γ' t₀) : γ = γ' :=
+  isIntegralCurve_eq_of_contMDiff
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/IntegralCurve.lean:550]: This would also fit on one line.

**Review by ocfnash — COMMENTED** [review]
I've left some minor comments but stepping back a bit I wonder if what you really want is the following lemma:
```lean
theorem ODE_solution_unique_of_mem_set_eventually {v : ℝ → E → E} {s : ℝ → Set E} {K : ℝ≥0}
    (hv : ∀ t, LipschitzOnWith K (v t) (s t))
    {f g : ℝ → E} {t₀ : ℝ}
    (hf : ∀ᶠ t in 𝓝 t₀, ContinuousAt f t)
    (hf' : ∀ᶠ t in 𝓝 t₀, HasDerivAt f (v t (f t)) t)
    (hfs : ∀ᶠ t in 𝓝 t₀, f t ∈ s t)
    (hg : ∀ᶠ t in 𝓝 t₀, ContinuousAt g t)
    (hg' : ∀ᶠ t in 𝓝 t₀, HasDerivAt g (v t (g t)) t)
    (hgs : ∀ᶠ t in 𝓝 t₀, g t ∈ s t)
    (ha : f t₀ = g t₀) : f =ᶠ[𝓝 t₀] g := by
  sorry
```
instead of `ODE_solution_unique_of_mem_set_Ioo` (and its variants) and then your main result would be:
```lean
theorem isIntegralCurveAt_eventuallyEq_of_contMDiffAt (hγt₀ : I.IsInteriorPoint (γ t₀))
    (hv : ContMDiffAt I I.tangent 1 (fun x ↦ (⟨x, v x⟩ : TangentBundle I M)) (γ t₀))
    (hγ : IsIntegralCurveAt γ v t₀) (hγ' : IsIntegralCurveAt γ' v t₀) (h : γ t₀ = γ' t₀) :
    γ =ᶠ[𝓝 t₀] γ' := by
```
Do you think this design would help?

File: Mathlib/Analysis/ODE/Gronwall.lean (hunk 4 above)
Comment (grunweg) [reply Mathlib/Analysis/ODE/Gronwall.lean:284]: Looking at it again, I think I referred to the have `hmt` on the next line.

(Two or three lines below, having to duplicate the same proof is not nice, but I'm not sure if extracting it would be that useful.)

**Comment by winstonyin** [comment]
Responding to your suggestion to switch to `Filter.Eventually`. YES, thank you. I learnt filters only halfway through this proof, and I'm fully convinced that it's much more user-friendly than the proliferation of `Ioo`, etc.

**Comment by winstonyin** [comment]
@ocfnash I simplified your suggested Grönwall statement a bit. See `isIntegralCurveAt_eventuallyEq_of_contMDiffAt` for the "eventually" version of integral curve uniqueness, where I've reduced code duplication by collecting some `have` statements.

**Review by ocfnash — APPROVED** [review]
I think the refactor to use `Filter.eventually` made a huge improvement. Thank you for all this great work.

Please apply the final suggestions and then feel free to merge!

bors d+
